beacon-chain/slasher/chunks.go

```diff
@@ -0,0 +1,548 @@
+// span chunks, this will be the start epoch of chunk index = (current chunk + 1). For example:
+//
+//                       chunk0     chunk1     chunk2
+//                         |          |          |
+//  max_spans_val_i = [[-, -, -], [-, -, -], [-, -, -]]
+//
+// If C = chunkSize is 3 epochs per chunk, and we input start epoch of chunk 1 which is 3. The next start
+// epoch is the start epoch of chunk 2, which is epoch 4. This is computed as:
+//
+//  first_epoch(chunkIndex(startEpoch)+1)
+//  first_epoch(chunkIndex(3)+1)
+//  first_epoch(1 + 1)
+//  first_epoch(2)
+//  4
+func (m *MaxSpanChunksSlice) NextChunkStartEpoch(startEpoch types.Epoch) types.Epoch {
+	return m.params.firstEpoch(m.params.chunkIndex(startEpoch) + 1)
+}
+
+// Given a validator index and epoch, retrieves the target epoch at its specific
+// index for the validator index and epoch in a min/max span chunk.
+func chunkDataAtEpoch(
+	params *Parameters, chunk []uint16, validatorIdx types.ValidatorIndex, epoch types.Epoch,
+) (types.Epoch, error) {
+	requiredLen := params.chunkSize * params.validatorChunkSize
+	if uint64(len(chunk)) != requiredLen {
+		return 0, fmt.Errorf("chunk has wrong length, %d, expected %d", len(chunk), requiredLen)
+	}
+	cellIdx := params.cellIndex(validatorIdx, epoch)
+	if cellIdx >= uint64(len(chunk)) {
+		return 0, fmt.Errorf("cell index %d out of bounds (len(chunk) = %d)", cellIdx, len(chunk))
+	}
+	distance := chunk[cellIdx]
+	return epoch.Add(uint64(distance)), nil
+}
+
+// Updates the value at a specific index in a chunk for a validator index + epoch
+// pair given a target epoch. Recall that for min spans, each element in a chunk
+// is the minimum distance between the a given epoch, e, and all attestation target epochs
+// a validator has created where att.source.epoch > e.
+func setChunkDataAtEpoch(
+	params *Parameters,
+	chunk []uint16,
+	validatorIdx types.ValidatorIndex,
+	epochInChunk,
+	targetEpoch types.Epoch,
+) error {
+	distance := epochDistance(targetEpoch, epochInChunk)
+	return setChunkRawDistance(params, chunk, validatorIdx, epochInChunk, distance)
+}
+
+// Updates the value at a specific index in a chunk for a validator index and epoch
+// to a specified, raw distance value.
+func setChunkRawDistance(
+	params *Parameters,
+	chunk []uint16,
+	validatorIdx types.ValidatorIndex,
+	epochInChunk types.Epoch,
+	distance uint16,
+) error {
+	cellIdx := params.cellIndex(validatorIdx, epochInChunk)
+	if cellIdx >= uint64(len(chunk)) {
+		return fmt.Errorf("cell index %d out of bounds (len(chunk) = %d)", cellIdx, len(chunk))
+	}
+	chunk[cellIdx] = distance
+	return nil
+}
+
+// Computes a distance between two epochs. Given the result stored in
+// min/max spans is at maximum WEAK_SUBJECTIVITY_PERIOD, we are guaranteed the
+// distance can be represented as a uint16 safely.
+func epochDistance(epoch, baseEpoch types.Epoch) uint16 {
+	return uint16(epoch.Sub(uint64(baseEpoch)))
+}
```

Comment:
One thing that could go wrong here is getting the arguments mixed up and underflow. Maybe check that `baseEpoch` is less than `epoch`, swap them if not. Sub will panic on underflow so you would probably know pretty quick if you got these backwards, i'm just thinking about how to make this more bullet proof.

Reply:
good idea